scripts/simplify_network.py

```diff
@@ -522,6 +523,151 @@ def consense(x):
     return clustering.network, clustering.busmap
 
 
+def remove_empty_isolated_nodes(n):
+    generators_sum_origin = n.generators.p_nom.sum()
+    load_sum_origin = n.loads_t.p_set.sum().sum()
+
+    n.determine_network_topology()
+
+    # duplicated sub-networks mean that there is at least one interconnection between buses
+    i_islands = n.buses[
+        (~n.buses.duplicated(subset=["sub_network"], keep=False))
+        & (n.buses.carrier == "AC")
+    ].index
+
+    # merge buses having sufficient attached load
+    loads_t_sum = n.loads_t.p_set.sum(axis=0)
+    load_islands = loads_t_sum.loc[loads_t_sum.index.isin(i_islands)]
+
+    # islands without generation and generation should be dropped
+    generation_islands = n.generators.loc[n.generators.bus.isin(i_islands)]
+    isl_no_load = set(i_islands) - set(load_islands.index)
+    isl_no_gen = set(i_islands) - set(generation_islands.bus)
+    isl_empty = set.intersection(isl_no_load, isl_no_gen)
+
+    if len(isl_empty) == 0:
+        return n
+
+    n.mremove("Bus", isl_empty)
+    n.determine_network_topology()
+
+    load_sum_final = n.loads_t.p_set.sum().sum()
+    generators_sum_final = n.generators.p_nom.sum()
+
+    logger.info(
+        f"Dropped {len(isl_empty)} buses. A resulted load discrepancy is {(100 * ((load_sum_final - load_sum_origin)/load_sum_origin)):2.1}% and {(100 * ((generators_sum_final - generators_sum_origin)/generators_sum_origin)):2.1}% for average load and generation capacity, respectivelly"
+    )
+
+    return n
+
+
+def drop_isolated_nodes(n, drop_e_threshold, n_periods_in_year=8760):
+    n.determine_network_topology()
+
+    generators_sum_origin = n.generators.p_nom.sum()
+    load_sum_origin = n.loads_t.p_set.sum().sum()
+
+    loads_t_ann = n_periods_in_year * n.loads_t.p_set.mean(axis=0)
+
+    # duplicated sub-networks mean that there is at least one interconnection between buses
+    i_islands = n.buses[
+        (~n.buses.duplicated(subset=["sub_network"], keep=False))
+        & (n.buses.carrier == "AC")
+    ].index
+
+    load_islands = loads_t_ann.loc[loads_t_ann.index.isin(i_islands)]
+    # drop_e_threshold, GWh
+    small_load_index = load_islands[load_islands <= (1e3 * drop_e_threshold)].index
```

Comment:
In general this block of code may be improved:
- load_sum_origin  does not account for the time length (if we move towards mean power that improves, only a mean is needed)
- In general, there is no need to compute loads_t_ann for all buses and then get the subset. you may do loads_t_ann[i_islands].mean() or alike.
- You make use of a lot of [].index. I think it can be cleaned. Interestingly, you may get the indices of those buses where demand is higher than the threshold and discard the difference.
  `load_islands = loads_t_ann.loc[loads_t_ann.index.intersection(i_islands)]`
  `small_load_index= i_islands.difference(load_islands[load_islands > threashold].index)`
  This approach should automatically discard those buses that do not have any load as load_islands won't contain in the index that bus
  
No need to check on generation: if an isolated bus does have generation but no load it is useless: no generation will be produced

Reply:
Thanks a lot! Revised